### ISSUE TITLE
Can't write to resources with negative priority

### ISSUE THREAD

**Issue by rfc2822** [siacs/Conversations#1571]
Conversations seems to always send to the bare JID without resource. So, you can't contact users who are online with negative priority (which is quite useful for mobile phones, because often you don't want the mobile phone as a "normal" messaging device, but only for messages which are especially sent to the mobile phone).

Edit: Message carbons are (by intention) not enabled on the server.

**Comment by philipflohr** [comment]
My experience with this is that Conversations sends to all ressources as long as there is no reason to don't do this (eg OTR).

The feature you're looking for is XEP-0168 http://xmpp.org/extensions/xep-0168.html
Personally I think the current behaviour is better;)

**Comment by d9h02f** [comment]
The idea is that a normal user doesn't want, doesn't know and shouldn't have to bother with such ridiculous nonsense as priorities, they just want their messages to be read by the other person.

**Comment by rfc2822** [comment]
Normal users just want to use WhatsApp and Facebook Messenger…

**Comment by escitalopram** [comment]
The user wouldn't need to know about about resources if they were implemented as recommended in [RFC 6121](https://tools.ietf.org/html/rfc6121#section-5.1).
If that procedure is followed, the only scenario when users need to know about resources is when they actually use them intentionally.

Therefore I'd rather think that many clients are feature crippled in this respect not because of the users' alleged lacking of intellectual capacity, but rather because there is virtually no software available that adheres to the ideals of interoperability practised in the early days of the internet, which are still embodied in the XMPP RFCs and most XEPs that managed to get their status past »experimental«.

Just imagine, if Outlook users could only read mails composed by other Outlook users and Thunderbird users could only read mails composed by other Thunderbird users. We wouldn't want that kind of situation, would we?

This is not about different Conversations clients talking to each other over Conversations-approved servers. It is about Conversations communicating with a multitude of XMPP services and clients as diverse as possible. The most important step on the way to achieve this goal is to have developers actually read, understand and implement the relevant specifications.

**Comment by strb** [comment]
> The user wouldn't need to know about about resources if they were implemented as recommended in RFC 6121.

Note that Conversations is NOT out of spec here. The RFC may RECOMMEND this behavior, but it doesn't mandate it. Quoting RFC 2119:

> [...] there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

There is, in fact a valid reason to deviate from the RECOMMENDED behavior.

> If that procedure is followed, the only scenario when users need to know about resources is when they actually use them intentionally.

This is simply not true. By not sending to full JIDs, we give users without Carbons support on their server a chance to achieve passable multi-device mobility. 

By setting a negative priority, you're implicitly agreeing to not receive messages that are not specifically targeted to that client. This means that under the RECOMMENDED behavior, in a scenario where you have a desktop and mobile client with mobile set to negative priority as outlined in the OP, once your desktop client goes offline, further messages sent by your contact will NOT reach your mobile device. They will either bounce or be queued in offline storage for delivery to the desktop, until you reply from your mobile device. 

This problem is exacerbated by the fact that in the mobile IM paradigm, connections might break down every now and again, even with things like stream management. Say you're on a train. Your desktop is offline, your phone is online with a negative priority. You are occasionally exchanging messages with a contact, and they're using what you would deem "correct" behavior by implementing the policy outlined in the RFC section you linked to. Now your connectivity drops for longer than 5 minutes (or whatever your stream management timeout may be). Your contacts will receive unavailable presences from your phone, which would "unlock" (in the terms of the RFC) them from sending to your phones full JID. Now you won't receive any messages from them on your phone until you send a message to them.

There are a ton of issues regarding the UX here, and it's not quite a cut-and-dried as you're presenting it. So basically, by setting a negative priority, you accept that messages may get misdelivered, and this issue is just about disagreement in the choice of WHEN this behavior should occur. It has never been the goal of Conversations to support every single obscure use-case out there. Conversations has always been about leveraging XMPP technology to provide a modern IM experience. We consider things synchronized history a must-have feature to achieve this goal, and we aim to provide the best experience possible in any given scenario.

> Just imagine, if Outlook users could only read mails composed by other Outlook users and Thunderbird users could only read mails composed by other Thunderbird users. We wouldn't want that kind of situation, would we?

I don't really see how this straw man argument for "standards serve a purpose" has anything to do with this discussion, as nobody is arguing otherwise. Just because people disagree with you about implementation choices (which are deliberately left somewhat open) doesn't mean they\re not adhering to the standard.

So in conclusion, this is not a use-case we can support properly without harming user experience, negative presence priorities are to be used with caution and they don't work exactly how you probably think they work, and this is not an issue isolated to Conversations.

**Comment by escitalopram** [comment]
It is people like YOU who are responsible for the nonexistent adoption of XMPP. If you want centralized history there are already XEPs for that (to lower myself to your level of discussion I'll not care to cite them, because you won't make use of them anyway) and that »Message Carbons« stillbirth won't help anything about it, it will just unnerve users.
YOU are actively hindering me (and many other people) from communicating effectively with my friends, even if I don't use your stupid android app, by ignoring carefully thought about specification recommendations that ensure basic interoperability with clients that actually support »modern« usage patterns (like multi-device, etc.)
Hope you are happy now.
